File: chart/code/symbology/src/symbolizer/symbolizer.cpp
#include "ogc/symbology/symbolizer/symbolizer.h"

namespace ogc {
namespace symbology {

struct Symbolizer::Impl {
    std::string name;
    ogc::draw::DrawStyle defaultStyle;
    bool enabled = true;
    double minScale = 0.0;
    double maxScale = std::numeric_limits<double>::max();
    int zIndex = 0;
    double opacity = 1.0;
};

Symbolizer::Symbolizer() : impl_(std::make_unique<Impl>()) {
}

Symbolizer::~Symbolizer() = default;

std::string Symbolizer::GetName() const {
    return impl_->name;
}

void Symbolizer::SetDefaultStyle(const ogc::draw::DrawStyle& style) {
    impl_->defaultStyle = style;
}

ogc::draw::DrawStyle Symbolizer::GetDefaultStyle() const {
    return impl_->defaultStyle;
}

void Symbolizer::SetEnabled(bool enabled) {
    impl_->enabled = enabled;
}

bool Symbolizer::IsEnabled() const {
    return impl_->enabled;
}

void Symbolizer::SetName(const std::string& name) {
    impl_->name = name;
}

void Symbolizer::SetMinScale(double scale) {
    impl_->minScale = scale;
}

double Symbolizer::GetMinScale() const {
    return impl_->minScale;
}

void Symbolizer::SetMaxScale(double scale) {
    impl_->maxScale = scale;
}

double Symbolizer::GetMaxScale() const {
    return impl_->maxScale;
}

bool Symbolizer::IsVisibleAtScale(double scale) const {
    if (scale <= 0) {
        return true;
    }
    return scale >= impl_->minScale && scale <= impl_->maxScale;
}

void Symbolizer::SetZIndex(int index) {
    impl_->zIndex = index;
}

int Symbolizer::GetZIndex() const {
    return impl_->zIndex;
}

void Symbolizer::SetOpacity(double opacity) {
    impl_->opacity = opacity;
}

double Symbolizer::GetOpacity() const {
    return impl_->opacity;
}

ogc::draw::DrawStyle Symbolizer::MergeStyle(const ogc::draw::DrawStyle& base, const ogc::draw::DrawStyle& override) const {
    ogc::draw::DrawStyle result = base;
    
    if (override.pen.IsVisible()) {
        result.pen = override.pen;
    } else if (base.pen.IsVisible()) {
        result.pen = base.pen;
    }
    
    if (override.brush.IsVisible()) {
        result.brush = override.brush;
    } else if (base.brush.IsVisible()) {
        result.brush = base.brush;
    }
    
    return result;
}

SymbolizerPtr Symbolizer::Create(SymbolizerType type) {
    (void)type;
    return nullptr;
}

std::string& Symbolizer::NameRef() {
    return impl_->name;
}

ogc::draw::DrawStyle& Symbolizer::DefaultStyleRef() {
    return impl_->defaultStyle;
}

bool& Symbolizer::EnabledRef() {
    return impl_->enabled;
}

double& Symbolizer::MinScaleRef() {
    return impl_->minScale;
}

double& Symbolizer::MaxScaleRef() {
    return impl_->maxScale;
}

int& Symbolizer::ZIndexRef() {
    return impl_->zIndex;
}

double& Symbolizer::OpacityRef() {
    return impl_->opacity;
}

}
}

File: chart/cycle/javawrapper/src/main/java/cn/cycle/chart/api/symbology/Symbolizer.java
package cn.cycle.chart.api.symbology;

import cn.cycle.chart.api.style.Color;
import cn.cycle.chart.api.style.DrawStyle;

public class Symbolizer {

    public enum Type {
        POINT,
        LINE,
        POLYGON,
        TEXT,
        RASTER
    }

    private final Type type;
    private String name;
    private DrawStyle style;
    private double minScale;
    private double maxScale;
    private String filter;
    private boolean enabled;

    public Symbolizer(Type type) {
        this.type = type;
        this.enabled = true;
        this.minScale = 0;
        this.maxScale = Double.MAX_VALUE;
    }

    public Type getType() {
        return type;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public DrawStyle getStyle() {
        return style;
    }

    public void setStyle(DrawStyle style) {
        this.style = style;
    }

    public double getMinScale() {
        return minScale;
    }

    public void setMinScale(double minScale) {
        this.minScale = minScale;
    }

    public double getMaxScale() {
        return maxScale;
    }

    public void setMaxScale(double maxScale) {
        this.maxScale = maxScale;
    }

    public String getFilter() {
        return filter;
    }

    public void setFilter(String filter) {
        this.filter = filter;
    }

    public boolean isEnabled() {
        return enabled;
    }

    public void setEnabled(boolean enabled) {
        this.enabled = enabled;
    }

    public boolean isScaleInRange(double scale) {
        return scale >= minScale && scale <= maxScale;
    }

    public static Symbolizer createPointSymbolizer() {
        Symbolizer s = new Symbolizer(Type.POINT);
        s.setName("Point");
        s.setStyle(DrawStyle.defaultPoint());
        return s;
    }

    public static Symbolizer createLineSymbolizer() {
        Symbolizer s = new Symbolizer(Type.LINE);
        s.setName("Line");
        s.setStyle(DrawStyle.defaultLine());
        return s;
    }

    public static Symbolizer createPolygonSymbolizer() {
        Symbolizer s = new Symbolizer(Type.POLYGON);
        s.setName("Polygon");
        s.setStyle(DrawStyle.defaultPolygon());
        return s;
    }

    public static Symbolizer createTextSymbolizer() {
        Symbolizer s = new Symbolizer(Type.TEXT);
        s.setName("Text");
        return s;
    }
}
